File: pa1/List.java
// Matthew Tan
// mxtan
// pa1: List.java
// This is a doubly-linked list ADT

public class List
{
    // private class Node
    private class Node
    {
        // Fields
        int data;
        Node prev;
        Node next;
      
        // Constructor
        public Node(int data) 
        { 
            this.data = data; 
            prev = null;
            next = null; 
        }
      
        // toString():  overrides Object's toString() method
        public String toString() 
        { 
            return String.valueOf(data); 
        }
      
        // equals(): overrides Object's equals() method
        public boolean equals(Object x)
        {
            boolean eq = false;
            Node that;
            if (x instanceof Node)
            {
                that = (Node) x;
                eq = (this.data == that.data);
            }
            return eq;
        }
    }

    // Fields
    private Node front;
    private Node back;
    private Node cursor;
    private int index;
    private int length;

    // Constructor
    public List() 
    {
        init();
    }
    
    // init() condenses all for List() constructor
    private void init()
    {
        front = null;
        back = null;
        cursor = null;
        index = -1;
        length = 0;
    }
        
    // -- Access functions --------------------------------------

    // length()
    // returns the number of elements in this List
    public int length()
    {
        return length;
    }
    
    // index()
    // if cursor is defined, returns the index of
    // the current element,
    // otherwise return -1
    public int index()
    {
        if (length() <= 0)
        {
            return -1;
        }
        return index;
    }

    // front()
    // returns front element. pre: length() > 0
    public int front()
    {
        if (length() <= 0)
        {
            throw new RuntimeException(
                  "ListException: front() called on empty list");
        }
        return front.data;
    }
     
    // back()
    // returns back element. pre: length() > 0
    public int back()
    {
        if (length() <= 0)
        {
            throw new RuntimeException(
                  "ListException: back() called on empty list");
        }
        return back.data;
    }
    
    // get()
    // returns the cursor element. pre: length() > 0 and index() > 0
    public int get()
    {
        if (length() <= 0)
        {
            throw new RuntimeException(
                  "ListException: cannot get() if empty");
        }
        if (cursor == null)
        {
            throw new RuntimeException(
                  "ListException: cannot get() if cursor is null");
        }
        return cursor.data;
    }

    // equals()
    // returns true if this List and L are the same integer
    // sequence. the cursor is ignored in both lists.
    public boolean equals(List L)
    {
        boolean eq = false;
        Node M, N;
        M = this.front;
        N = L.front;
        eq = (this.length() == L.length());
        while (eq && M != null)
        {
            eq = M.equals(N);
            M = M.next;
            N = N.next;
        }
        return eq;
    }

    // -- Mainpulation procedure --------------------------
    
    // clear()
    // rests this List to it original empty state
    public void clear()
    {
        init();

        // alternate clear:
        /*
        front = null;
        back = null;
        cursor = null;
        index = -1;
        length = 0;
        */
    }

    // moveFront()
    // if List is non-empty, places the cursor under
    // the front element, otherwise does nothing
    public void moveFront()
    {
        if (length() > 0)
        {
            cursor = front;
            index = 0;
        }
    }

    // moveBack()
    // if list is non-empty, places the cursor under
    // the back element, otherwise does nothing
    public void moveBack()
    {
        if (length() > 0)
        {
            cursor = back;
            index = length - 1;
       }
    }
    
    // movePrev()
    // if cursor is defined and not at front, moves cursor
    // one step toward front of this List, cursor
    // becomes undefined, if cursor is undefined does nothing
    public void movePrev()
    {
        if (cursor == front)
        {
            cursor = null;
            index = -1;
        }
        else
        {
            cursor = cursor.prev;
            index--;
        }
    }

    // moveNext
    // if cursor is defined and not at front, move cursor
    // one step toward back of this List, if cursor is defined,
    // and at front, cursor becomes undefined, if
    // cursor is undefined, does nothing
    public void moveNext()
    {
        if (cursor == back)
        {
            cursor = null;
            index = -1;
        }
        else
        {
            cursor = cursor.next;
            index++;
        }
    }

    // prepend()
    // insert new elment into this List. if List is non-empty,
    // insertion takes place before front element
    public void prepend(int data)
    {
        Node N = new Node(data);
        if (front == null) 
        {
            front = N;
            back = N;
            cursor = front;
        } 
        else 
        {
            front.prev = N;
            N.next = front;
            N.prev = null;
            front = N;
            index++;
        }
        length++;
    }
    
    // append()
    // insert new element into this List. if this lis is non-empty
    // insertion takes place after back element
    public void append(int data)
    {
        Node N = new Node(data);
        if (back == null) 
        {
            front = N;
            back = N;
            cursor = back;
        } 
        else 
        {
            back.next = N;
            N.prev = back;
            back = N;
            N.next = null;
        }
        length++;
    }
    
    // insertBefore()
    // insert new element before cursor
    // pre: length() > 0, index() > 0
    public void insertBefore(int data)
    {
        if (length() <= 0)
        {
            //throw new RuntimeException(
            //      "ListException: cannot insertBefore() in an empty list");
            prepend(data);
        }
        if (cursor == null)
        {
            throw new RuntimeException(
             "ListException: cannot insertBefore() if cursor is undefined");
        }
        if (cursor == front)
        {
            prepend(data);
        }
        else
        {
            Node N = new Node(data);
            //System.out.println("in insertBefore(): Node N is: " + N);
            //System.out.println("in insertBefore(): cursor is: " + cursor);
            N.prev = cursor.prev;
            //System.out.println("in insertBefore(): N.prev is: " + N.prev);
            //System.out.println("in insertBefore(): cursor.prev is: " + cursor.prev);
            N.next = cursor;
            //System.out.println("in insertBefore(): N.next is: " + N.next);
            cursor.prev.next = N;
            cursor.prev = N;
            index++;
            length++;
        }
    }
    
    // insertAfter()
    // insert new element after cursor
    // pre: length() > 0, index() > 0
    public void insertAfter(int data)
    {
        if (length() <= 0)
        {
            throw new RuntimeException(
                 "ListException: insertAfter() called on empty List");
        }
        if (cursor == back) 
        {
            append(data);
        }
        else 
        {
            Node N = new Node(data);
            cursor.next.prev = N;
            N.next = cursor.next;
            N.prev = cursor;
            cursor.next = N;
            length++;
        }
    }
    
    // deleteFront()
    // deletes front element. pre: length() > 0
    public void deleteFront()
    {
        if (length() <= 0)
        {
            throw new RuntimeException(
                "ListException: cannot deleteFront() empty list.");
        }
        else
        {
            if (front == cursor)
            {
                cursor = null;
                index = -1;
            }
            if (length > 1)
            {
                front = front.next;
                front.prev = null;
                if (cursor != null)
                {                    
                    index--;
                }
            }
            length--;
        }
    }

    // deleteBack()
    // deletes back element. pre: length() > 0
    public void deleteBack()
    {
        if (length() <= 0)
        {
            throw new RuntimeException(
                  "ListException :cannot deleteBack() on an empty list");
        }
        
        else
        {
            if (back == cursor)
            {
                //System.out.println("deleteBack() called: " + "cursor is: " + cursor);
                cursor = null;
                //System.out.println("deleteBack() called: " + "cursor is: " + cursor);
                index = -1;
            }
            //System.out.println("deleteBack(): resetting back: ");
            if (length > 1)
            {
                back = back.prev;
                back.next = null;
            }
            length--;
        }
    }
    
    // delete()
    // deletes cursor element, making cursor undefined
    // pre: length() > 0, index() > 0
    public void delete()
    {
        //System.out.println("List: delete() called");
        if (length() <= 0)
        {
            throw new RuntimeException(
                  "ListException: cannot delete() empty list");
        }
        if (cursor == null)
        {
            throw new RuntimeException(
                  "ListException: cannot delete() on undefined cursor");
        }
        if (cursor == front)
        {
            deleteFront();
        }
        else if (cursor == back)
        {
            //System.out.println("deleteBack() called");
            deleteBack();
        }
        else
        {
            //System.out.println("in delete: cursor is: " + cursor);
            cursor.prev.next = cursor.next;
            cursor.next.prev = cursor.prev;
            length--;
        }
        index = -1;
    }

    // -- Other methods --------------------------------------

    // toString()
    // overrides object's toString method. returns a string
    // representation of this List containing a space
    public String toString()
    {
        StringBuffer sb = new StringBuffer();
        Node N = front;
        while (N != null)
        {
            sb.append(N.toString());
            if (N.next != null)
            {
                sb.append(" ");
            }
            N = N.next;
        }
        return new String(sb);
    }

    // copy()
    // returns a new List representing the same integer
    // sequence as this List. the cursor in the new lis
    // is undefined, regardless of the state of the cursor
    // in this list. the List is unchanged
    public List copy()
    {
        List newList = new List();
        if (length() > 0)
        {
            Node N = front;
            while (N != null)
            {
                newList.append(N.data);
                N = N.next;
            }
        }
        newList.cursor = null;
        newList.index = -1;
        return newList;
    }
}

File: pa4/List.c
// Matthew Tan
// mxtan
// pa4
// List.c: Doubly-linked list ADT written in C

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// struct is same as public class Node in java
typedef struct NodeObj
{
    int data;
    struct NodeObj* prev;
    struct NodeObj* next;
}NodeObj;

typedef NodeObj* Node; // private Node type

// struct is same as public class List in java
typedef struct ListObj
{
    Node front;
    Node back;
    Node cursor;
    int index;
    int length;
}ListObj;

// newNode()
// same as Node class constructor
Node newNode(int data)
{
    Node N = NULL;
    N = malloc(sizeof(NodeObj));
    //printf("List.c: newNode(): N malloc address: %p\n", N);
    N->data = data;
    N->prev = NULL;
    N->next = NULL;
    return (N);
}

// freeNode()
// frees heap memory pointed to by *pN, sets *pN to NULL
void freeNode(Node* pN)
{
    //printf("List.c: freeNode(): pN is: %p\n", pN);
    //printf("List.c: freeNode(): *pN is: %p\n", *pN);
    if (pN != NULL && *pN != NULL)
    {
        //printf("!!List.c: freeNode(): free(*pN) called\n");
        free(*pN);
        *pN = NULL;
    }
}

// newList()
// equivalent to List class constructor
List newList(void)
{
    List L = NULL;
    L = malloc(sizeof(ListObj));
    //printf("List.c: newList(): L malloc address: %p\n", L);
    L->front = NULL;
    L->back = NULL;
    L->cursor = NULL;
    L->index = -1;
    L->length = 0;
    return (L);
}

// freeList()
// frees all heap memory associated with List *pL
// and sets *pL to NULL
void freeList(List* pL)
{
    //printf("List.c: freeList() called\n");
    if (pL != NULL && *pL!= NULL) 
    { 
      while (length(*pL) > 0)
      {            
          //printf("List.c: freeList(): deleteFront() called\n");
          deleteFront(*pL);
      }
      //printf("!!List.c: freeList(): free(*pL) called for List\n");
      free(*pL);
      *pL = NULL;
    }
}

// Access functions ---------------------------------------

// length()
// returns the number of elements in this List
int length(List L)
{
    if (L == NULL)
    {
        printf("List Error: calling length() on NULL List reference\n");
        exit(1);
    }
    return L->length;
}

// index()
// returns the index of the cursor of the list
int index(List L)
{
    //printf("List.c: index() is called.\n");
    if (L == NULL)
    {
        printf("List Error: calling index() on NULL List reference\n");
        exit(1);
    }
    if (length(L) <= 0)
    {
        //printf("List.c: index() called: got to if statement.\n");
        return -1;
    }
    //printf("List.c: index(): index is: %d\n", L->index);
    return L->index;
}

// front()
// returns the front element. pre: length() > 0
int front(List L)
{
    if (L == NULL)
    {
        printf("List Error: calling front() on NULL List reference\n");
        exit(1);
    }
    if (length(L) <= 0)
    {
        printf("List Error: front() called on empty List\n");
    }
    return L->front->data;
}

// back()
// returns the back element. pre: length() > 0
int back(List L)
{
    if (L == NULL)
    {
        printf("List Error: calling back() on NULL List reference\n");
        exit(1);
    }
    if (length(L) <= 0)
    {
        printf("List Error: back() called on empty List\n");
    }
    return L->back->data;
}

// get()
// returns the cursor element. pre: length() > 0 and index() > 0
int get(List L)
{
    if (L == NULL)
    {
        printf("List Error: calling get() on NULL List reference\n");
        exit(1);
    }
    if (length(L) <= 0)
    {
        printf("List Error: get() called on empty List\n");
    }
    if (L->cursor == NULL)
    {
        printf("List Error: cannot get() if cursor is NULL\n");
    }
    return L->cursor->data;
}

// equals
// returns 1 if this List and L are the 
// same integer sequence.
//the cursor is ignored in both lists.
int equals(List A, List B)
{
    int eq = 0;
    Node M = NULL;
    Node N = NULL;
    if (A == NULL || B == NULL)
    {
        printf("List Error: calling equals() on NULL List reference\n");
        exit(1);
    }
    eq = (A->length == B->length);
    M = A->front;
    N = B->front;
    while (eq && M != NULL)
    {
        eq = (M->data == N->data);
        M = M->next;
        N = N->next;  
    }
    return eq;
}

// Manipulation procedures ------------------------------------------------

// clear()
// resets this List to its original empty state
void clear(List L)
{
    if (L == NULL)
    {
        printf("List Error: calling clear() on a NULL List reference\n");
        exit(1);
    }
    while (length(L) > 0)
    {            
        //printf("List.c: freeList(): deleteFront() called\n");
        deleteFront(L);
    }
    L->front = NULL;
    L->back = NULL;
    L->cursor = NULL;
    L->index = -1;
    L->length = 0;
}

// moveFront()
// places the cursor to the front of the List
void moveFront(List L)
{
    //printf("List.c: in moveFront(): length is: %d\n", length(L));
    if (L == NULL)
    {
        printf("List Error: calling moveFront() on a NULL List reference\n");
        exit(1);
    }
    if (length(L) > 0)
    {
        L->cursor = L->front;
        L->index = 0;
    }
}

// moveBack()
// places the cursor at the back of the List
void moveBack(List L)
{
    if (L == NULL)
    {
        printf("List Error: calling moveBack() on a NULL List reference\n");
        exit(1);
    }
    if (length(L) > 0)
    {
        L->cursor = L->back;
        L->index = L->length - 1;
    }
}

// movePrev()
// moves cursor one step toward the front of the List
void movePrev(List L)
{
    if (L == NULL)
    {
        printf("List Error: calling movePrev() on a NULL List reference\n");
        exit(1);
    }
    if (L->cursor == L->front)
    {
        L->cursor = NULL;
        L->index = -1;
    }
    else
    {
        L->cursor = L->cursor->prev;
        L->index--;
    }
}

// moveNext()
// moves the cursor one step toward the back of the List
void moveNext(List L)
{
    if (L == NULL)
    {
        printf("List Error: calling moveNext() on a NULL List reference\n");
        exit(1);
    }
    if (L->cursor == L->back)
    {
        L->cursor = NULL;
        L->index = -1;
    }
    else
    {
        L->cursor = L->cursor->next;
        L->index++;
    }
}

// prepend()
// inserts a new element into the front of the List
void prepend(List L, int data)
{
    if (L == NULL)
    {
        printf("List Error: calling prepend() on a NULL List reference\n");
        exit(1);
    }
    Node N = newNode(data);
    if (L->front == NULL)
    {
        L->front = N;
        L->back = N;
        L->cursor = L->front;
    }
    else
    {
        L->front->prev = N;
        N->next = L->front;
        L->front = N;
        L->index++;
    }
    L->length++;
}

// append()
// inserts a new element into the back of the List
void append(List L, int data)
{
    if (L == NULL)
    {
        printf("List Error: calling apend() on a NULL List reference\n");
        exit(1);
    }
    Node N = newNode(data);
    if (L->back == NULL)
    {
        L->front = N;
        L->back = N;
        L->cursor = L->back;
    }
    else
    {
        L->back->next = N;
        N->prev = L->back;
        L->back = N;
        N->next = NULL;
    }
    L->length++;
}

// insertBefore()
// inserts new element before cursor
void insertBefore(List L, int data)
{
    if (L == NULL)
    {
        printf("List Error: calling apend() on a NULL List reference\n");
        exit(1);
    }
    if (L->cursor == NULL)
    {
        printf("List Error: cannot insertBefore()  if cursor is undefined\n");
    }
    if (L->cursor == L->front)
    {
        prepend(L, data);
    }
    else
    {
        Node N = newNode(data);
        N->prev = L->cursor->prev;
        N->next = L->cursor;
        L->cursor->prev->next = N;
        L->cursor->prev = N;
        L->index++;
        L->length++;
    }
}

// insertAfter()
// inserts new element after cursor
void insertAfter(List L, int data)
{
    if (L == NULL)
    {
        printf("List Error: calling apend() on a NULL List reference\n");
        exit(1);
    }
    if (length(L) <= 0)
    {
        printf("List Error: insertAfter() called on empty List\n");
    }
    if (L->cursor == L->back)
    {
        append(L, data);
    }
    else
    {
        Node N = newNode(data);
        L->cursor->next->prev = N;
        N->next = L->cursor->next;
        N->prev = L->cursor;
        L->cursor->next = N;
        L->length++;
    }
}

// deleteFront()
// deletes front element
void deleteFront(List L)
{
    //printf("List.c: deleteFront() called\n");
    if (L == NULL)
    {
        printf("List Error: calling apend() on a NULL List reference\n");
        exit(1);
    }
    if (length(L) <= 0)
    {
        printf("List Error: cannot deleteFront() empty List\n");
    }
    else
    {
        //printf("List.c: freeList(): node is: %d\n", L->front->data);    

        //printf("---------------------------------------\n");
        //printf("List.c: deleteFront(): L->length is: %d\n", L->length);
        //printf("List.c: deleteFront(): L->front is: %p\n", L->front);
        //printf("List.c: deleteFront(): L->cursor is: %p\n", L->cursor);

      if (L->length == 1)
      {
          Node N = L->front;
          //printf("List.c: case 1: deleteFront(): N is: %p\n", N);
          freeNode(&N);
          //printf("List.c: case 1: deleteFront(): front.data is: %d\n", L->front->data);
          L->cursor = NULL;
          L->front = L->back = NULL;
          L->index = -1;
      }
      else //if (L->length > 1)
      {
          Node N = L->front;
          //printf("List.c: case 2: deleteFront(): front.data is: %d\n", L->front->data);
          L->front = L->front->next;
          L->front->prev = NULL;
          if (L->cursor != NULL)
          {
              L->index--;
          }
          freeNode(&N);
      }
      L->length--;
  }
}

// deleteBack()
// deletes back element
void deleteBack(List L)
{
    if (L == NULL)
    {
        printf("List Error: cannot deleteBack() on NULL List reference\n");
    }
    if (length(L) <= 0)
    {
        printf("List Error: cannot deleteBack() empty List\n");
    }
    else
    {
        if (L->length == 1)
        {
            Node N = L->back;
            freeNode(&N);
            L->cursor = NULL;
            L->front = L->back = NULL;
            L->index = -1;
        }
        else //if (L->length > 1)
        {
            Node N = L->back;
            L->back = L->back->prev;
            L->back->next = NULL;
            if (L->index == L->length - 1)
            {
                L->index = -1;
            }
            freeNode(&N);
        }
        L->length--;
        //printf("List.c: deleteBack(): current index is: %d\n", L->index);
    }
}

// delete()
// deletes cursor element
void delete(List L)
{
    if (L == NULL)
    {
        printf("List Error: cannot deleteBack() on NULL List reference\n");
    }
    if (length(L) <= 0)
    {
        printf("List Error: cannot delete() empty List");
    }
    if (L->cursor == NULL)
    {
        printf("List Error: cannot delete() on undefined cursor");
    }
    if (L->cursor == L->front)
    {
        deleteFront(L);
    }
    else if (L->cursor == L->back)
    {
        deleteBack(L);
    }
    else
    {
        Node N = L->cursor;
        L->cursor->prev->next = L->cursor->next;
        L->cursor->next->prev = L->cursor->prev;
        freeNode(&N);
        L->length--;
    }
    L->index = -1;
}

// printList()
// prints contents to a file
void printList(FILE* out, List L) 
{
    Node N = NULL;
    if(L == NULL) 
    {
        printf("List error: printList() called on NULL List reference\n");
        exit(1);
    }
    for (N = L->front; N != NULL; N = N->next) 
    {
        fprintf(out, "%d", N->data);
        fprintf(out, " ");
    }
}

// copyList()
// makes a copty of the List
List copyList(List L)
{
    List C = newList();
    if (length(L) > 0)
    {
        Node N = L->front;
        while (N != NULL)
        {
            append(C, N->data);
            N = N->next;
        }
    }
    C->cursor = NULL;
    C->index = -1;
    return C;
}
